UefiTestingPkg/FunctionalSystemTests/MemoryAttributeProtocolFuncTestApp/AARCH64/PageSplitTest.c

```diff
@@ -0,0 +1,122 @@
+/** @file -- PagingAuditProcessor.c
+
+Platform specific memory audit functions.
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+// MdePkg is the ideal place for this test, but because the AARCH64 PageSplitTest relies on
+// ArmLib to get the translation table base address and TCR, this test is in the UefiTestingPkg to avoid
+// circular repo dependencies in Project Mu.
+#include <Library/ArmLib.h>
+#include "MemoryAttributeProtocolFuncTestApp.h"
+
+#define TT_ADDRESS_MASK  (0xFFFFFFFFFULL << 12)
+
+#define IS_TABLE(page, level)  ((level == 3) ? FALSE : (((page) & TT_TYPE_MASK) == TT_TYPE_TABLE_ENTRY))
+#define IS_BLOCK(page, level)  ((level == 3) ? (((page) & TT_TYPE_MASK) == TT_TYPE_BLOCK_ENTRY_LEVEL3) : ((page & TT_TYPE_MASK) == TT_TYPE_BLOCK_ENTRY))
+#define ROOT_TABLE_LEN(T0SZ)   (TT_ENTRY_COUNT >> ((T0SZ) - 16) % 9)
+
+/**
+  Get an un-split page table entry and allocate entire region so the page
+  doesn't need to be split on allocation
+
+  @param[out]  Address  Address of allocated 2MB page region
+**/
+EFI_STATUS
+EFIAPI
+GetUnsplitPageTableEntry (
+  OUT EFI_PHYSICAL_ADDRESS  *Address
+  )
+{
+  UINT64                *Pml0          = NULL;
+  UINT64                *Pte1G         = NULL;
+  UINT64                *Pte2M         = NULL;
+  UINT64                Index2         = 0;
+  UINT64                Index1         = 0;
+  UINT64                Index0         = 0;
+  UINT64                RootEntryCount = 0;
+  EFI_STATUS            Status;
+  EFI_PHYSICAL_ADDRESS  BaseAddress;
+
+  Pml0 = (UINT64 *)ArmGetTTBR0BaseAddress ();
+
+  if (Pml0 == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  RootEntryCount = ROOT_TABLE_LEN (ArmGetTCR () & TCR_T0SZ_MASK);
+
+  for (Index0 = 0x0; Index0 < RootEntryCount; Index0++) {
+    if (!IS_TABLE (Pml0[Index0], 0)) {
+      continue;
+    }
+
+    Pte1G = (UINT64 *)(Pml0[Index0] & TT_ADDRESS_MASK);

```

Comment:
Because the translation table is architecturally defined and there is no recovery scenario for a malformed table, letting the system blow up here seems fine to me. Certainly when the CPU does these table walks it isn't validating the sanity of the table.